### PR TITLE
Add GoogleAdsClient to DI container

### PR DESCRIPTION
**Pull request by Raibaz**
This replicates #540 and rebases it on top of `main` to allow merging.

### CODE REVIEW

**Review by sundquist — CHANGES_REQUESTED** [review]

File: Google.Ads.GoogleAds.Core/src/Lib/GoogleAdsClient.cs [hunk 1]
```diff
@@ -12,20 +12,44 @@
...
 using Google.Ads.GoogleAds.Config;
 using Google.Api.Gax.Grpc;
 using Grpc.Core.Interceptors;
-using System;
 using System.Collections.Generic;
 
 namespace Google.Ads.GoogleAds.Lib
 {
+    /// <summary>
+    /// The Google Ads Client interface to allow for DI container configuration.
```
[the remark this answers is not in the document]
Comment (Raibaz) [reply Google.Ads.GoogleAds.Core/src/Lib/GoogleAdsClient.cs]: I think it's a fairly common acronym, but I don't see a reason not to spell it out entirely.

Changed it.

File: Google.Ads.GoogleAds.Core/src/Lib/GoogleAdsClient.cs [hunk 2]
```diff
@@ -12,20 +12,44 @@
...
+    /// </summary>
+    public interface IGoogleAdsClient
+    {
+        /// <summary>
+        /// <para>Gets an instance of the specified service. Use this method with a predefined
+        /// list of templates available for each supported version. E.g.</para>
```
[the remark this answers is not in the document]
Comment (Raibaz) [reply Google.Ads.GoogleAds.Core/src/Lib/GoogleAdsClient.cs]: Done.

File: Google.Ads.GoogleAds.Extensions.DependencyInjection/src/GoogleAdsClientServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        /// </summary>
+        public static IServiceCollection AddGoogleAdsClient(this IServiceCollection services, IConfiguration config)
+        {
+            IConfigurationSection googleAdsSection = config.GetSection(GoogleAdsApiOptions.GoogleAdsApi);
+            if (!googleAdsSection.Exists())
+            {
+                throw new Exception($"Missing {GoogleAdsApiOptions.GoogleAdsApi} configuration.");
+            }
+
+            // add the setting to the DI container as options
```
[the remark this answers is not in the document]
Comment (Raibaz) [reply Google.Ads.GoogleAds.Extensions.DependencyInjection/src/GoogleAdsClientServiceCollectionExtensions.cs]: Good point, made it consistent with the rest of the codebase with the capital letter and the period. Also spelt out dependency injection while I was there.

File: Google.Ads.GoogleAds.Core/src/Lib/GoogleAdsClient.cs [hunk 4]
```diff
@@ -12,20 +12,44 @@
...
+        /// <para>
+        /// // Gets an instance of V10.GoogleAdsServiceClient<br/>
+        /// GoogleAdsServiceClient googleAdsService = client.GetService(Services.V10.GoogleAdsService);
+        /// </para>
+        /// </summary>
+        /// <param name="serviceTemplate"><para>The template that corresponds to the service to
+        /// be created. Predefined service templates are available for each supported version, as
+        /// a nested class within the <code>Services</code> class.</para>
+        /// <para>E.g. <code>Services.V10</code> class contains predefined templates for v10 of
```
[the remark this answers is not in the document]
Comment (Raibaz) [reply Google.Ads.GoogleAds.Core/src/Lib/GoogleAdsClient.cs]: Done

**Review by sundquist — APPROVED** [review]
LGTM

**Review by AnashOommen — APPROVED** [review]

**Comment by sommmen** [comment]
Hiya quick question - i must be blind or something, but i just updated the ads package and now i get the following error;

``` 
System.AggregateException
  HResult=0x80131500
  Message=One or more errors occurred. (Status(StatusCode="Internal", Detail="Error starting gRPC call. TokenResponseException: Error:"invalid_request", Description:"Missing required parameter: refresh_token", Uri:""", DebugException="Google.Apis.Auth.OAuth2.Responses.TokenResponseException: Error:"invalid_request", Description:"Missing required parameter: refresh_token", Uri:"""))
  Source=Google.Ads.Gax
  StackTrace:
   at Google.Ads.Gax.Interceptors.UnaryRpcInterceptor.<>c__DisplayClass1_0`2.<Intercept>b__0(Task`1 oldTask)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Google.Api.Gax.Grpc.ApiCall.GrpcCallAdapter`2.<>c__DisplayClass4_0.<<CallAsync>g__WaitAndCallHandlers|0>d.MoveNext()
   at Google.Api.Gax.Grpc.ApiCallRetryExtensions.<>c__DisplayClass0_0`2.<<WithRetry>b__0>d.MoveNext()
   at OPG.Jobs.Ads.GoogleAdsSyncJob.<OnRun>d__3.MoveNext() in C:\Repos\dotnet-api\OPG.Jobs\Ads\GoogleAdsSyncJob.cs:line 54

  This exception was originally thrown at this call stack:
    Google.Apis.Auth.OAuth2.Responses.TokenResponse.FromHttpResponseAsync(System.Net.Http.HttpResponseMessage, Google.Apis.Util.IClock, Google.Apis.Logging.ILogger)
    Google.Apis.Auth.OAuth2.Requests.RequestExtensions.PostFormAsync(object, System.Net.Http.HttpClient, string, System.Net.Http.Headers.AuthenticationHeaderValue, Google.Apis.Util.IClock, Google.Apis.Logging.ILogger, System.Threading.CancellationToken)
    Google.Apis.Auth.OAuth2.Flows.AuthorizationCodeFlow.FetchTokenAsync(string, Google.Apis.Auth.OAuth2.Requests.TokenRequest, System.Threading.CancellationToken)
    Google.Apis.Auth.OAuth2.Flows.AuthorizationCodeFlow.FetchTokenAsync(string, Google.Apis.Auth.OAuth2.Requests.TokenRequest, System.Threading.CancellationToken)
    Google.Apis.Auth.OAuth2.Flows.AuthorizationCodeFlow.RefreshTokenAsync(string, string, System.Threading.CancellationToken)
    Google.Apis.Auth.OAuth2.UserCredential.RefreshTokenAsync(System.Threading.CancellationToken)
    Google.Apis.Auth.OAuth2.TokenRefreshManager.RefreshTokenAsync()
    Google.Apis.Auth.OAuth2.TokenRefreshManager.GetAccessTokenForRequestAsync(System.Threading.CancellationToken)
    Google.Apis.Auth.OAuth2.UserCredential.GetAccessTokenWithHeadersForRequestAsync(string, System.Threading.CancellationToken)
    Grpc.Auth.GoogleAuthInterceptors.FromCredential.AnonymousMethod__0(Grpc.Core.AuthInterceptorContext, Grpc.Core.Metadata)
    ...
    [Call Stack Truncated]

Inner Exception 1:
RpcException: Status(StatusCode="Internal", Detail="Error starting gRPC call. TokenResponseException: Error:"invalid_request", Description:"Missing required parameter: refresh_token", Uri:""", DebugException="Google.Apis.Auth.OAuth2.Responses.TokenResponseException: Error:"invalid_request", Description:"Missing required parameter: refresh_token", Uri:""")

Inner Exception 2:
TokenResponseException: Error:"invalid_request", Description:"Missing required parameter: refresh_token", Uri:""

```

Startup.cs
`services.AddSingleton<GoogleAdsClient>();`

GoogleAdsSyncJob.cs
``` csharp
var customerService = _client.GetService(Services.V17.CustomerService);

var listAccessibleCustomersResponse = await customerService
    .ListAccessibleCustomersAsync(new ListAccessibleCustomersRequest(), CallSettings.FromCancellationToken(cancellationToken));
```

I then see in the release notes that auth can be added to DI now as mentioned in this PR.
So i try calling `AddGoogleAdsClient()` as mentioned in the tests and it failes to find the method.

The nuget browser also cannot see the extension;
image

I also tried v18 instead of v17, same issue.

**Comment by Raibaz** [comment]
Ah, you're right, the Google.Ads.GoogleAds.Extensions package wasn't updated in the last release of the client library.

We'll be publishing another release soon which will include these changes as well.

Sorry for the inconvenience.

**Pull request merged**